portal-impl/src/com/liferay/portlet/PortletAsyncListenerAdapter.java

```diff
@@ -54,30 +54,40 @@ public void addListener(
 			ResourceRequest resourceRequest, ResourceResponse resourceResponse)
 		throws IllegalStateException {
 
-		try {
-			if (_firedOnError) {
+		if (_firedOnError) {
+			try {
 				portletAsyncListener.onError(
 					new PortletAsyncEvent(
 						_portletAsyncContext, resourceRequest,
 						resourceResponse));
 			}
+			catch (IOException ioe) {
+				_log.error("Failed to notify listener for onError.", ioe);
+			}
+		}
 
-			if (_firedOnTimeout) {
+		if (_firedOnTimeout) {
+			try {
 				portletAsyncListener.onTimeout(
 					new PortletAsyncEvent(
 						_portletAsyncContext, resourceRequest,
 						resourceResponse));
 			}
+			catch (IOException ioe) {
+				_log.error("Failed to notify listener for onTimeout.", ioe);
+			}
+		}
 
-			if (_firedOnComplete) {
+		if (_firedOnComplete) {
+			try {
 				portletAsyncListener.onComplete(
 					new PortletAsyncEvent(
 						_portletAsyncContext, resourceRequest,
 						resourceResponse));
 			}
-		}
-		catch (IOException ioe) {
-			_log.error(ioe, ioe);
+			catch (IOException ioe) {
```

Comment:
@tinatian 
Compared to last PR, we made changes to follow Tomcat's behavior for notification-after-completed:

1. Use separate try-catches to make sure logging is clear and exception from `onError` or `onTimeout` doesn't prevent `onComplete` to be invoked;
2. Adjust the order to make sure `onComplete` is invoked after `onError` or `onTimeout`, if the complete state is a result of error or timeout.

This feature is not required by the Spec, but we think it's better to have, since most of other frameworks work this way.